66_HLSLBxDFTests/app_resources/tests.hlsl

```diff
@@ -0,0 +1,455 @@
+> {};
+
+template<class T>
+struct is_microfacet_brdf : bool_constant<
+    is_same<T, bxdf::reflection::SBeckmannBxDF<sample_t, iso_cache, aniso_cache>>::value ||
+    is_same<T, bxdf::reflection::SGGXBxDF<sample_t, iso_cache, aniso_cache>>::value
+> {};
+
+template<class T>
+struct is_basic_bsdf : bool_constant<
+    is_same<T, bxdf::transmission::SLambertianBxDF<sample_t, iso_interaction, aniso_interaction>>::value
+> {};
+
+template<class T>
+struct is_microfacet_bsdf : bool_constant<
+    is_same<T, bxdf::transmission::SBeckmannDielectricBxDF<sample_t, iso_cache, aniso_cache>>::value ||
+    is_same<T, bxdf::transmission::SGGXDielectricBxDF<sample_t, iso_cache, aniso_cache>>::value
+> {};
+
+template<class T>
+NBL_CONSTEXPR bool is_basic_brdf_v = is_basic_brdf<T>::value;
+template<class T>
+NBL_CONSTEXPR bool is_microfacet_brdf_v = is_microfacet_brdf<T>::value;
+template<class T>
+NBL_CONSTEXPR bool is_basic_bsdf_v = is_basic_bsdf<T>::value;
+template<class T>
+NBL_CONSTEXPR bool is_microfacet_bsdf_v = is_microfacet_bsdf<T>::value;
+
+
+template<class BxDF, bool aniso = false>
+struct TestUOffset : TestBxDF<BxDF>
+{
+    using base_t = TestBase<BxDF>;
+    using this_t = TestUOffset<BxDF, aniso>;
+
+    float32_t4 test()
+    {
+        sample_t s, sx, sy;
+        quotient_pdf_t pdf;
+        float32_t3 brdf;
+        aniso_cache cache, dummy;
+
+        if NBL_CONSTEXPR_FUNC (is_basic_brdf_v<BxDF>)
+        {
+            s = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u.xy);
+            sx = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u.xy + float32_t2(base_t::rc.h,0));
+            sy = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u.xy + float32_t2(0,base_t::rc.h));
+        }
+        if NBL_CONSTEXPR_FUNC (is_microfacet_brdf_v<BxDF>)
+        {
+            s = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u.xy, cache);
+            sx = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u.xy + float32_t2(base_t::rc.h,0), dummy);
+            sy = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u.xy + float32_t2(0,base_t::rc.h), dummy);
+        }
+        if NBL_CONSTEXPR_FUNC (is_basic_bsdf_v<BxDF>)
+        {
+            s = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u);
+            sx = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u + float32_t3(base_t::rc.h,0,0));
+            sy = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u + float32_t3(0,base_t::rc.h,0));
+        }
+        if NBL_CONSTEXPR_FUNC (is_microfacet_bsdf_v<BxDF>)
+        {
+            s = base_t::bxdf.generate(base_t::anisointer, base_t::rc.u, cache);
+            float32_t3 ux = base_t::rc.u + float32_t3(base_t::rc.h,0,0);
+            sx = base_t::bxdf.generate(base_t::anisointer, ux, dummy);
+            float32_t3 uy = base_t::rc.u + float32_t3(0,base_t::rc.h,0);
+            sy = base_t::bxdf.generate(base_t::anisointer, uy, dummy);
+        }
```

Comment:
`h` should be a parameter of each individual specialized BxDF test, because its specific to the differential test and also may need to change depending on the BxDF getting tested

Reply:
also maybe compute the perturbed `u` above for readability